### PR TITLE
treehouses usb [on|off] added (fixes #538)

### LINKED ISSUE #538: a treehouses cmd for turn on|off usb ports?
https://webworxshop.com/switching-raspberry-pi-usb-power-via-home-assistant-and-octoprint/

### PR DESCRIPTION
**Pull request by kjong**
Implements #538 

A feature to turn the usb ports on or off (cannot control individual usb ports due to raspberry pi usb hub design).

`./cli.sh usb on`
`./cli.sh usb off`

Use latest image (117+) when testing.

Tested on Raspberry Pi 4 and 3B+

**KNOWN ISSUES**
* ~Turning off the usb ports while the raspberry pi is connected via ethernet will mess up your ethernet connection (usb ports + ethernet are connected)~ Added a check to see if ethernet is connected
* After doing some reading, it seems that you can disable data through the usbs but not power them off completely, which would explain things like usb fans still spinning after "turning off" the usb ports
* Possible example of a usb fan being powered on / off using `uhubctl` on a rpi 3B+ (http://www.byfarthersteps.com/?p=6802)

**WIP**
* ~Add checks for hub-ctrl binary~
* ~Add checks for libusb-dev pkg (will use new `check_missing_packages` from #577)~

### CODE REVIEW

**Comment by hiroTochigi** [comment]
uhubctl works on rpi3 b+.

image

**Comment by kjong** [comment]
http://download.treehouses.io/experiment/treehouse-usb0.img.gz

Use this image when testing

**Comment by Nida8930** [comment]
Tested using treehouse-usb0 and treehouses-116 images on a rpi4B and rpi3B+.

`./cli.sh usb on` and `./cli.sh usb off` worked correctly on both images using the rpi4B. 

The usb feature initially did not work on the rpi3B+ with either image. 

After installing hub-ctrl and its dependancies I was able to turn the usb ports on and off by running the command `./hub-ctrl -h 1 -P 2 -p 0`. If I ran `./cli.sh usb off` or `./cli.sh usb on` then the previous hub-ctrl command would no longer work.  

After further investigation I found that booting up the pi and running `lsusb` would list the 4 usb ports and anything I had plugged in. After running `./cli.sh usb off` this list is reduced to a single usb port and running `./cli.sh usb on' did not return the list to how it was before running the off command. Rebooting the pi was the only way I found to get the pi to recognize all four usb ports again. 

Screen Shot 2019-12-31 at 8 11 43 PM


When using `./hub-ctrl -h 1 -P 2 -p 0` and `./hub-ctrl -h 1 -P 2 -p 1` the pi continually recognized all four usb ports with the device I had plugged in absent when the usb ports are powered off.

Screen Shot 2019-12-31 at 8 15 12 PM

I think that adding a check at the beginning of the script to determine the model of the pi and adjusting the commands accordingly could get the usb.sh script to work on the rpi3 and rpi4.

**Comment by kjong** [comment]
Thanks for testing @Nida8930, checks for what rpi the user is running is in the works, after I sort out issues with builder and the new image.

**Comment by LordJashin32** [comment]
@irisb1701 @hiroTochigi @Nida8930 
** What is that cool screenshot utility you're using to grab console? Gimmie please. **
related:
img

**Comment by kjong** [comment]
Added rpi3 functionality courtesy of @hiroTochigi

File: modules/usb.sh [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  if [ ! -e /usr/local/bin/hub-ctrl ]; then
+    echo "required binary 'hub-ctrl' not found"
+    exit
+  fi
+
+  # check if libusb-dev pkg is installed
+  if [ "$(dpkg-query -W -f='${Status}' libusb-dev 2>/dev/null | grep -c 'ok installed')" -eq 0 ]; then
+    echo "required package 'libusb-dev' not installed"
+    exit
+  fi
```
Comment (dogi) [inline modules/usb.sh]: https://github.com/treehouses/cli/blob/master/modules/globals.sh#L144-L157

**Comment by jamjakpa** [comment]
Does it still to be tested on PI 3 ?

**Comment by jamjakpa** [comment]
> Implements #538
> 
> A feature to turn the usb ports on or off (cannot control individual usb ports due to raspberry pi usb hub design).
> 
> `./cli.sh usb on`
> `./cli.sh usb off`
> 
> http://download.treehouses.io/experiment/treehouse-usb2.img.gz
> 
> Use this image when testing.
> 
> Tested on Raspberry Pi 4.
> Please test on other rpi models if possible.
> 
> **KNOWN ISSUES**
> 
> * ~Turning off the usb ports while the raspberry pi is connected via ethernet will mess up your ethernet connection (usb ports + ethernet are connected)~ Added a check to see if ethernet is connected
> * Won't function if libusb package is missing (requires `sudo apt install libusb-dev build-essential git`)
> * After doing some reading, it seems that you can disable data through the usbs but not power them off completely, which would explain things like usb fans still spinning after "turning off" the usb ports
> * Possible example of a usb fan being powered on / off using `uhubctl` on a rpi 3B+ (http://www.byfarthersteps.com/?p=6802)
> 
> **WIP**
> 
> * Test to see if usb ports can be controlled by hubs (ie. turn off hub 1 usb ports while leaving hub 2 usb ports on)
> * Add checks for libusb-dev pkg and hub-ctrl binary

So i flash that image http://download.treehouses.io/experiment/treehouse-usb2.img.gz
it doesn't boot up on my Raspberry Pi 3 Model B+ . 
I just see how its loading for like 3 seconds then its breaking up. 
Does anyone else had that issues that its not loading the image ?

**Pull request merged**